### ISSUE TITLE
Terraform vSphere provider Windows support

### ISSUE THREAD

**Issue by Supernomad** [hashicorp/terraform#5490]
So I realized a bit too late that Windows is just not supported on the current version of the vSphere provider.

However it seems like it could be with proper use of `custom_configuration_parameters`, but I can't seem to figure out what the syntax should be. I have taken a few different routes to no avail and was curious if anyone had some insight.

I have tried the following:
```
custom_configuration_parameters = {
  spec.identity.guiUnattended.autoLogon = "true"
  spec.identity.guiUnattended.autoLogonCount = 1
  spec.identity.guiUnattended.password.plainText = "true"
  <-- SNIP -->
}
```

```
custom_configuration_parameters = {
  identity.guiUnattended.autoLogon = "true"
  identity.guiUnattended.autoLogonCount = 1
  identity.guiUnattended.password.plainText = "true"
  <-- SNIP -->
}
```

```
custom_configuration_parameters = {
  spec = {
    identity = {
      guiUnattended = {
        autoLogon = "true"
        autoLogonCount = 1
        password.plainText = "true"
      }
    }
  }
    <-- SNIP -->
}
```

**Comment by rossedman** [comment]
@phinze @Supernomad I really need help on this. The fact this isn't supported just tanked me half way through an engagement. This isn't stated in the docs at all. Does anyone have any idea on how to make this work.

**Comment by pimpajr** [comment]
I'm attempting this with some Windows and Linux vm's as well and noticing the same thing. As long as the key doesn't have any dots in it the matching value will be populated. As soon as you add in dots the value gets dropped. Admittedly I've really only done this kind of thing with ruby, but I attempted to troubleshoot some. Inserting a debug line before [Line 338 in resource_vsphere_virtual_machine.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/vsphere/resource_vsphere_virtual_machine.go#L338) printing out vL the values are dropped by this point. I'll be honest I don't have enough experience in Go yet to know why this is happening or to come up with a fix myself. I believe this part is related to [Issue 5933](https://github.com/hashicorp/terraform/issues/5933). The [test for custom_configuration_parameters](https://github.com/hashicorp/terraform/blob/master/builtin/providers/vsphere/resource_vsphere_virtual_machine_test.go#L627) looks to only be testing keys with no dots as well. 

More related to windows specifically it looks like the [provider is using CustomizationLinuxPrep](https://github.com/hashicorp/terraform/blob/master/builtin/providers/vsphere/resource_vsphere_virtual_machine.go#L1184) and doesn't have a differentiation between Windows and Linux in order to call CustomizationSysprep. 

This is about as far as I've gotten trying to get this to work short of passing no custom parameters since all the ones I need have dots in them. I know this doesn't solve the issue, but hopefully it helps anyone with more Go experience than I come up with a fix.

**Comment by rossedman** [comment]
@pimpajr I hadn't noticed the CustomizationLinuxPrep. I have a branch I am working on where I allowed the `guest_id` to be changed in the resource because Linux is hardcoded [here](https://github.com/hashicorp/terraform/blob/master/builtin/providers/vsphere/resource_vsphere_virtual_machine.go#L910) as well. Do you know much about the CustomizationLinuxPrep stuff?

**Comment by rossedman** [comment]
@pimpajr I bet this fails because it falls back to Linux prep and then inserts the wrong name for WindowsPrep.

**Comment by pimpajr** [comment]
@rossedman good catch on the guest_id, I missed that one. I'm fairly familiar with CustomizationLinuxPrep in ruby, I usually just check [vmware's reference here](http://pubs.vmware.com/vsphere-60/index.jsp#com.vmware.wssdk.apiref.doc/vim.vm.customization.LinuxPrep.html). It's pretty limited for Linux, it's well represented in provider go code actually. [CustomizationSysprep](http://pubs.vmware.com/vsphere-60/index.jsp#com.vmware.wssdk.apiref.doc/vim.vm.customization.Sysprep.html) is bit different in that it takes [CustomizationUserData](http://pubs.vmware.com/vsphere-60/index.jsp#com.vmware.wssdk.apiref.doc/vim.vm.customization.UserData.html) and [CustomizationIdentification](http://pubs.vmware.com/vsphere-60/index.jsp#com.vmware.wssdk.apiref.doc/vim.vm.customization.Identification.html) where your typical sysprep settings like joinDomain, can be set.

**Comment by pimpajr** [comment]
@rossedman I'm not really sure how to implement any of what I mentioned above in Go so if there's any specific information about the vsphere api I can help with I'd be happy to.

**Comment by rossedman** [comment]
@pimpajr Yeah, I am trying to build Terraform using these objects but it isn't recognizing them. Do you have any good resources on learning about some of these deeper prep stuff? Would really like to solve this even if its partial. I know VMWare's guest customizations are pretty deep

**Comment by rossedman** [comment]
@pimpajr This look like its literally just nesting the objects that the vSphere API uses. Here is an example:

Linux:
```
		Identity: &types.CustomizationLinuxPrep{
			HostName: &types.CustomizationFixedName{
				Name: strings.Split(vm.name, ".")[0],
			},
			Domain:     vm.domain,
			TimeZone:   vm.timeZone,
			HwClockUTC: types.NewBool(true),
		},
```

Windows:
```
		Identity: &types.CustomizationSysprep{
			Identification: types.CustomizationIdentification{
				DomainAdmin: "Administrator",
			},
		},
```

**Comment by pimpajr** [comment]
@rossedman I see what you're saying now. Best I can find is https://raw.githubusercontent.com/vmware/govmomi/master/vim25/types/types.go and search for CustomizationSysprep.

**Comment by rossedman** [comment]
@pimpajr I got a different error running this. Made it past the spec.identity stuff so I am guessing I am on the right track just need to provide all required properties.

**Comment by clstokes** [comment]
I believe the _dot_ issue is the same as https://github.com/hashicorp/terraform/issues/2143.

**Comment by aheeren** [comment]
Can I get some feedback on https://github.com/hashicorp/terraform/pull/6087 from interested parties? WOMM.

**Comment by rossedman** [comment]
@aheeren You are a saint. This looks much better than what I was working on. I will pull yours and build and test as well

**Comment by stack72** [comment]
Closed via #6087
